[ATO-706] Add documentation for secrets managers (#12110)
RasaHQ/rasa: docs/docs/secrets-managers.mdx

---
id: secrets-managers

> sidebar_label: Secrets Managers

Author: My 2 cents:
I originally named to feature Credentials Management but in the code I used Secrets Manager a lot. Mostly because I worked with AWS Secrets Manager in the past.
From code standpoint it would be easier to use Secrets Manager and to rename discovery ticket in JIRA.

@andreaklmn WDYT?

> sidebar_label: Secrets Managers

Reviewer: I think we should name it in a way that makes sense for our customers and our overall messaging. If both are equivalent, I agree with what you said and we should pick the simpler choice for us. But only if they're equivalent :D

> sidebar_label: Secrets Managers

Reviewer 2: I guess for me the question is: are we managing all types of sensitive data, including passwords, API keys, encryption keys, certificates, and other types of confidential information - if yes, then it's "secrets management"

If we are only managing credentials, I would probably stick to credentials management..? Unless we are planning on managing other "secrets" as part of this solution

My 2 cents!

> sidebar_label: Secrets Managers

Reviewer 2: I just saw that the Vault website here uses the 2 terms somewhat interchangeably: https://www.vaultproject.io/
Under "automate credentials rotation" - so I'm down for using secrets management!

> sidebar_label: Secrets Managers

Author: For anyone stumbling on this PR: We have discussed this and agreed to use Secrets Manager.
Just to put my answer here as well, we are able to manage sensitive data.

title: Secrets Managers
description: Credentials used to authenticate to external services are safely stored in a secrets manager.

> abstract: Credentials managers are used to safeguard credentials used to authenticate to external services.

Reviewer: I would rephrase to:

> You can store your assistant's secrets in an external credentials manager. Rasa Pro currently supports credentials manager for the Tracker Store (link to the tracker store doc page).

---

<!-- this file is version specific, do not use `@site/...` syntax -->
import RasaProLabel from '@theme/RasaProLabel';
import RasaProBanner from '@theme/RasaProBanner';

<RasaProLabel />
<RasaProBanner/>

## Secrets Managers

> Secrets managers are used to safeguard credentials used to authenticate to external services,

Reviewer 3: As a general rule of thumbs, avoid passive voice as much as possible:
```suggestion
Secrets managers safeguard credentials used to authenticate access to external services,
```

like databases.

Rasa Pro supports the following secrets managers:
* [HashiCorp Vault](https://www.hashicorp.com/products/vault)

> Currently, Rasa Pro supports sefeguarding credentials for the following services:

Reviewer: ```suggestion
Currently, Rasa Pro supports safeguarding credentials for the following services:
```

- [Tracker Stores](./tracker-stores.mdx)

> ## Vault Secrets Manager

Reviewer: ```suggestion
## HashiCorp Vault Secrets Manager
```

Reviewer: I think we need consistent naming for this one too

> The Vault Secrets Manager is used to store credentials used to authenticate to external services.

Reviewer 3: ```suggestion
The Vault Secrets Manager is used to store credentials used to authenticate access to external services.
```

> The credentials are stored in a Vault instance and can be encrypted at rest.

Reviewer 3: We should rephrase to active voice and provide a link to Vault documentation/tutorial on how to store credentials in a Vault instance, e.g. `To store credentials in a Vault instance, you can read the official Vault docs. You can also encrypt credentials at rest.`

Reviewer 3: Perfect 👍🏻

If you need to encrypt your credentials,
you can use the [Vault Transit Engine](https://www.vaultproject.io/docs/secrets/transit).

:::note

> Expiring tokens need to be renewed periodically, and renewal process is done over the network,

Reviewer: ```suggestion
Expiring tokens need to be renewed periodically, and the renewal process is done over the network,
```

> we will try to renew the token 5 seconds before it expires.

Reviewer: ```suggestion
Rasa Pro will try to renew the token 5 seconds before it expires.
```

> If token's TTL is less than 5 seconds, we will try to renew it ASAP,

Reviewer: ```suggestion
If the token's time-to-live (TTL) is less than 5 seconds, we will try to renew it as soon as possible,
```

Reviewer: what does "as soon as possible" mean in this context?

Author: Now I have realized that we have not covered this use case in the best way. 
If token's TTL is less than 5 seconds we should renew the token right away instead of waiting for token's remaining TTL.

Author: I opened [PR](https://github.com/RasaHQ/rasa-plus/pull/365) to address this situation.

Author: After aforementioned PR is merged, we can say that it will be renewed after 1 second.

> but it might fail due to network latency.

Reviewer: I think we need a suggestion / recommendation here. What happens if it fails, what will users need to do? Do we recommend TTL > 5s ? What does HashiCorp recommend?

Author: > What happens if it fails, what will users need to do?
Users will have to create a new token and restart Rasa with that new token.

We can build re-try mechanism for renewal of tokens.

HashiCorp does not have general recommendation for Vault token's TTL.
We can say that we recommend that token's TTL to be at least 1 minute.

Author: I opened [PR](https://github.com/RasaHQ/rasa-plus/pull/372) to address this.

To accommodate for network latency I increased threshold for renewal so that it starts 15 seconds before token will expire and retry 5 times every second if renewal fails.

:::

### Authentication

> Rasa Pro supports authentication to Vault through
> [Token](https://www.vaultproject.io/docs/auth/token).

Reviewer: ```suggestion
Rasa Pro supports connections to Vault via
[Token authentication](https://www.vaultproject.io/docs/auth/token).
```

Author: I would recommend we use:

```
Rasa Pro can authenticate to Vault through
[Token authentication](https://www.vaultproject.io/docs/auth/token).
````

because Rasa will authenticate to Vault through token. To say that it connects to Vault through token authentication sounds a bit mixed to me.

Both `expiring` and `non-expiring`(so called, root tokens) tokens are supported.
Rasa Pro will automatically renew the token if it is expiring.

### How to configure access to Vault

Access to Vault secrets manager can be configured with environment variables
and through `endpoints.yml` configuration file.

Environment variables and `endpoints.yml` configuration file are merged together

> and *the values from the environment variables take precedence*.

Reviewer: nice -- would be great to know if this is a pattern that we strictly follow in other Rasa features cc @madebymagnolia

> Following environment variables are available:

Reviewer: ```suggestion
The following environment variables are available:
```

> ```bash

Reviewer 4: | Environment Variable          | Description                                                                                    | Default               |
| :---------------------------- | :----------------------------------------------------------------------------------------------| :-------------------- |
| `SECRET_MANAGER`              | **Required**. The secrets manager to use. *Currently only "vault" is supported*                | `vault`               |
| `VAULT_HOST`                  | **Required**. The address of the vault server                                                  |                       |
| `VAULT_TOKEN`                 | **Required**. token to authenticate to the vault server                                        |                       |
| `VAULT_RASA_SECRETS_PATH`     | Path to the secrets in the vault server                                                        | `rasa-secrets`        |
| `VAULT_TRANSIT_MOUNT_POINT`   | If transit secrets engine is enabled, set this to  mount point of the transit engine           |                       |

Author: To keep env vars and endpoints.yml consistent WDYT about:

Toe following environment variables are available:
| Environment Variable          | Description                                                                                    | Default               |
| :---------------------------- | :----------------------------------------------------------------------------------------------| :-------------------- |
| `SECRET_MANAGER`              | **Required**. The secrets manager to use. *Currently only "vault" is supported*                | `vault`               |
| `VAULT_HOST`                  | **Required**. The address of the vault server                                                  |                       |
| `VAULT_TOKEN`                 | **Required**. token to authenticate to the vault server                                        |                       |
| `VAULT_RASA_SECRETS_PATH`     | Path to the secrets in the vault server                                                        | `rasa-secrets`        |
| `VAULT_TRANSIT_MOUNT_POINT`   | If transit secrets engine is enabled, set this to mount point of the transit engine            |                       |

To configure the Vault secrets manager, you can fill the following section in `endpoints.yml` file:
```yaml-rasa title="endpoints.yml
secrets_manager:
    type: vault # required - the secrets manager to use
    token: <token used to authenticate to Vault> # required - token to authenticate to the vault server   
    url: "http://localhost:1234" # required - the address of the vault server      
    secrets_path: rasa-secrets  # path to the secrets in the vault server if not set it defaults to `rasa-secrets`
    transit_mount_point: transit # if transit secrets engine is enabled, set this to mount point of the transit engine
```

Author: @Balowen Could you take a look at my comment above?

Reviewer 5: Looks good now

> SECRET_MANAGER = "vault" # which secrets manager to use, currently only vault is supported
> VAULT_HOST # address of the vault server
> VAULT_TOKEN # token to authenticate to the vault server
> VAULT_RASA_SECRETS_PATH # path to the secrets in the vault server, if not set it defaults to `rasa-secrets`
> VAULT_TRANSIT_MOUNT_POINT # if transit secrets engine is enabled, set this to  mount point of the transit engine

Reviewer: ```suggestion
SECRET_MANAGER="vault" # which secrets manager to use, currently only 'vault' is supported
VAULT_HOST="<host>" # address of the vault server
VAULT_TOKEN="<token>" # token to authenticate to the vault server
VAULT_RASA_SECRETS_PATH="<path>" # path to the secrets in the vault server, if not set it defaults to `rasa-secrets`
VAULT_TRANSIT_MOUNT_POINT="<mount-point>" # if transit secrets engine is enabled, set this to mount point of the transit engine
```

Author: Addressed in [comment](https://github.com/RasaHQ/rasa/pull/12110#discussion_r1130770521)

```

> The `endpoints.yml` configuration for vault secrets manager looks like this:

Reviewer 3: ```suggestion
To configure the Vault secrets manager, you can fill the following section in `endpoints.yml` file:
```

> ```yaml-rasa {3} title="endpoints.yml

Reviewer: is the `{3}` highlight intended? I don't think so 🤔

> secrets_manager:

Reviewer: I would add exactly the same comments. If users look at the doc, they might just look at the yml file doc if this is their preferred way, and will have to scroll up to look at the doc for each variable. (I'm talking about comments like "# which secrets manager to use, currently only 'vault' is supported")

    type: vault
    token: <token used to authenticate to Vault>

>     url: "http://localhost:1234"

Reviewer: `url` here and `VAULT_HOST` above are kind of confusing. Can't we unify the variable names? (pick url or host)

Author: Opened [PR](https://github.com/RasaHQ/rasa-plus/pull/366) to address change in the code

    secrets_path: rasa-secrets  # if not set it defaults to `rasa-secrets`
    transit_mount_point: transit # if you have enabled transit secrets engine, and you want to use it
```

> #### How to combine environment variables and `endpoints.yml`

Reviewer: I think the wording here needs to be clearer and say that we recommend storing sensitive information in the environment variables instead of files. WDYT?

Author: How does:
```
####  Store access credentials in environment variables
```
sound?

A simple example on how to combine environment variables and `endpoints.yml` configuration file
would be to store access token in the environment variable and the rest of the configuration
in the `endpoints.yml` file.

```bash
# environment variables

> VAULT_TOKEN=<token used to authenticate to Vault>

Reviewer: ```suggestion
VAULT_TOKEN="<token>" # token to authenticate to the vault server
```
(consistent with what's written above)

```

> ```yaml-rasa {3} title="endpoints.yml

Reviewer: is the `{3}` highlight intended? I don't think so 🤔

secrets_manager:
    type: vault
    url: "http://localhost:1234"
    secrets_path: rasa-secrets  # if not set it defaults to `rasa-secrets`
    transit_mount_point: transit # if you have enabled transit secrets engine, and you want to use it
```

> ### Example

Reviewer: I would rework this section and add 2 sections:
- one about Configuring the tracker store and showing how to write the yml file  (line 130)
- one about an Example -> I would check with CSE if the example you wrote down is good enough for our customers

Author: @Balowen @l2nguyen What do you think about the example written here?
Should we keep it/improve it/remove it?

Author: > I would rework this section and add 2 sections:
> 
>     * one about Configuring the tracker store and showing how to write the yml file  (line 130)

I can add:

### How to configure Tracker Store to use Vault Secrets Manager
1. Configure Rasa to access the Vault instance

    Checkout the [How to configure access to Vault](#how-to-configure-access-to-vault) section for more details.

2. Configure Rasa to use the Vault secrets manager to fetch credentials for the tracker store
    ```yaml-rasa title="endpoints.yml
            tracker_store:
                  type: SQL
                  url: localhost:5432
                  username:
                        source: secrets_manager.vault
                        secret_key: sql_store_username
                  password:
                        source: secrets_manager.vault
                        secret_key: sql_store_password
    ```

WDYT?

Author: I have remove the example section.

> This is a simple example which demonstrates how to store credentials for tracker store in Vault

Reviewer: ```suggestion
This is a simple example which demonstrates how to store credentials for the Tracker Store in Vault
```

and how to access them from Rasa.

1. Start Vault instance
    ```bash
    docker run --cap-add=IPC_LOCK --name=vault -p 1234:1234 -p 8200:8200 -e 'VAULT_DEV_ROOT_TOKEN_ID=myroot' -e 'VAULT_DEV_LISTEN_ADDRESS=0.0.0.0:1234' -d vault
    ```

2. Store credentials for tracker store in Vault
    ```bash
    # go inside Vault container
    docker exec -it vault sh

    # (optional) if you get access denied error
    export VAULT_ADDR=http://127.0.0.1:1234
    export VAULT_TOKEN=myroot

    # store secret in the Vault under the group rasa-secrets
    vault kv put -mount=secret rasa-secrets sql_store_username=postgres sql_store_password=12345
    ```

3. Start Postgres container
    ```bash
    docker run --name postgres -e POSTGRES_DB=rasa -e POSTGRES_PASSWORD=12345 -p 5432:5432 -d postgres
    ```

4. Configure Rasa to access the Vault instance

    Checkout the [How to configure access to Vault](#how-to-configure-access-to-vault) section for more details.

5. Configure Rasa to use the Vault secrets manager to fetch credentials for the tracker store

>     ```yaml-rasa {3} title="endpoints.yml

Reviewer: is the `{3}` highlight intended? I don't think so 🤔

        tracker_store:
              type: SQL
              url: localhost:5432
              username:
                    source: secrets_manager.vault
                    secret_key: sql_store_username
              password:
                    source: secrets_manager.vault
                    secret_key: sql_store_password
    ```

> 6. Start Rasa

Reviewer 5: What do we mean here by `Start Rasa`? Does it mean Rasa server needs to be shut down when adding secrets to the HashiCorp Vault? Will the changes in endpoint.yml get applied if the user restarts rasa pod? I'm thinking here about our customers that would want to implement HashiCorp Vault into their current deployment

Reviewer 5: We should be more specific here, this example is essential to set up the Vault and use it properly.

Author: Rasa does not need to be stopped when secrets to Vault are added.
Rasa only needs to be stopped when we are adding/updating Vault connection configuration as Rasa currently does not support hot-reloading of it's configuration.

Author: @Balowen 
I have updated the example to be more comprehensive.
Let me know what do you think.

Reviewer 5: I think it's better now but I also agree with @l2nguyen [here](https://github.com/RasaHQ/rasa/pull/12110#discussion_r1132194604). It's unlikely our customers will ever run it on local.

---
id: secrets-managers
sidebar_label: Secrets Managers
title: Secrets Managers
description: Safeguard credentials your service uses to authenticate to external resources.
abstract: You can store your assistant's secrets in an external credentials manager.
    Rasa Pro currently supports credentials manager for the Tracker Store
---

<!-- this file is version specific, do not use `@site/...` syntax -->
import RasaProLabel from '@theme/RasaProLabel';
import RasaProBanner from '@theme/RasaProBanner';

<RasaProLabel />
<RasaProBanner/>

Rasa Pro supports the following secrets managers:
* [HashiCorp Vault](https://www.hashicorp.com/products/vault)

Currently, Rasa Pro supports safeguarding credentials for the following services:
- [Tracker Stores](./tracker-stores.mdx)

## HashiCorp Vault Secrets Manager

Use Vault Secrets Manager to store credentials used to authenticate access to external services.
The credentials are stored in a Vault instance and can be encrypted at rest.
To store credentials in a Vault instance, you can read the official Vault docs
[Storing secrets in Vault](https://developer.hashicorp.com/vault/tutorials/getting-started/getting-started-first-secret).

> You can also encrypt credentials at rest.
> you can use the [Vault Transit Engine](https://www.vaultproject.io/docs/secrets/transit).

Reviewer 5: ```suggestion
You can also encrypt credentials at rest.
You can use the [Vault Transit Engine](https://www.vaultproject.io/docs/secrets/transit).
```
Not sure if this should be two separate sentences, or one. If one then we should rephrase it somehow.

Author: @Balowen 
I updated this to 
```
You can also encrypt credentials at rest with [Vault Transit Engine](https://www.vaultproject.io/docs/secrets/transit).
```
WDYT?

Reviewer 5: LGTM

:::note
Expiring tokens need to be renewed periodically, and the renewal process is done over the network,
Rasa Pro will try to renew the token 15 seconds before it expires.
If the token's time-to-live (TTL) is less than 15 seconds, we will try to renew it after 1 second,
but it might fail due to network latency.

Rasa Pro has a built-in retry mechanism for renewing the token.

If the token is not renewed successfully it will be considered expired
and Rasa Pro will not be able to access the secrets.
You will need to create a new renewable token and restart Rasa Pro with new token.
:::

### Authentication
Rasa Pro can authenticate to Vault through
[Token authentication](https://www.vaultproject.io/docs/auth/token).

> Both `expiring` and `non-expiring`(so called, root tokens) tokens are supported.

Reviewer 5: ```suggestion
Both `expiring` and `non-expiring` (so called, root tokens) tokens are supported.
```

          url: localhost:5432
          username:
                source: secrets_manager.vault
                secret_key: sql_store_username
          password:
                source: secrets_manager.vault
                secret_key: sql_store_password
    ```

### Example
This is a simple example which demonstrates how to store credentials for the Tracker store in Vault
and how to access them from Rasa.

1. Start Vault instance
    ```bash
    docker run --cap-add=IPC_LOCK --name=vault -p 1234:1234 -p 8200:8200 -e 'VAULT_DEV_ROOT_TOKEN_ID=myroot' -e 'VAULT_DEV_LISTEN_ADDRESS=0.0.0.0:1234' -d vault
    ```

2. Store credentials for tracker store in Vault
    ```bash
    # go inside Vault container
    docker exec -it vault sh

    # (optional) if you get access denied error
    export VAULT_ADDR=http://127.0.0.1:1234
    export VAULT_TOKEN=myroot

    # store secret in the Vault under the group rasa-secrets
    vault kv put -mount=secret rasa-secrets sql_store_username=postgres sql_store_password=12345
    ```

3. Start Postgres container
    ```bash
    docker run --name postgres -e POSTGRES_DB=rasa -e POSTGRES_PASSWORD=12345 -p 5432:5432 -d postgres
    ```

> 4. [Configure tracker store with vault secerts manager](#how-to-configure-tracker-store-with-vault-secrets-manager)

Reviewer 5: ```suggestion
4. [Configure tracker store with vault secrets manager](#how-to-configure-tracker-store-with-vault-secrets-manager)
```

---
id: secrets-managers
sidebar_label: Secrets Managers
title: Secrets Managers
description: Safeguard credentials your service uses to authenticate to external resources.
abstract: You can store your assistant's secrets in an external credentials manager.
    Rasa Pro currently supports credentials manager for the Tracker Store [Tracker Stores](./tracker-stores.mdx).
---

<!-- this file is version specific, do not use `@site/...` syntax -->
import RasaProLabel from '@theme/RasaProLabel';
import RasaProBanner from '@theme/RasaProBanner';

<RasaProLabel />
<RasaProBanner/>

> ## Secrets Managers

Author: I will remove the title and sentence below it.
We should list which secret managers we support and in which service (with links for easier navigation).

    transit_mount_point: transit # if transit secrets engine is enabled, set this to mount point of the transit engine
```

#### Store access credentials in environment variables
A simple example on how to combine environment variables and `endpoints.yml` configuration file
would be to store access token in the environment variable and the rest of the configuration
in the `endpoints.yml` file.

```bash
# environment variables
VAULT_TOKEN=<token used to authenticate to Vault>
```

```yaml-rasa title="endpoints.yml
secrets_manager:
    type: vault
    url: "http://localhost:1234"
    secrets_path: rasa-secrets  # if not set it defaults to `rasa-secrets`
    transit_mount_point: transit # if you have enabled transit secrets engine, and you want to use it
```

### How to configure Tracker Store to read authentication credentials from Vault Secrets Manager
1. Configure Rasa to access the Vault instance

    Checkout the [How to configure access to Vault](#how-to-configure-access-to-vault) section for more details.

2. Configure Rasa to use the Vault secrets manager to fetch credentials for the tracker store
    ```yaml-rasa title="endpoints.yml

>             tracker_store:

Reviewer: top level indentation probably needs to be reset

    ```

2. Store credentials for tracker store in Vault
    ```bash
    # go inside Vault container
    docker exec -it vault sh

    # (optional) if you get access denied error
    export VAULT_ADDR=http://127.0.0.1:1234
    export VAULT_TOKEN=myroot

    # store secret in the Vault under the group rasa-secrets
    vault kv put -mount=secret rasa-secrets sql_store_username=postgres sql_store_password=12345
    ```

3. Start Postgres container
    ```bash
    docker run --name postgres -e POSTGRES_DB=rasa -e POSTGRES_PASSWORD=12345 -p 5432:5432 -d postgres
    ```

4. Configure Rasa to access the Vault instance

    Checkout the [How to configure access to Vault](#how-to-configure-access-to-vault) section for more details.

5. Configure Rasa to use the Vault secrets manager to fetch credentials for the tracker store
    ```yaml-rasa title="endpoints.yml
        tracker_store:
              type: SQL
              url: localhost:5432
              username:
                    source: secrets_manager.vault
                    secret_key: sql_store_username
              password:
                    source: secrets_manager.vault
                    secret_key: sql_store_password

>     ```

Reviewer: this duplicates the content above, I would remove and link to it
